src/LineSegment.cpp

```diff
@@ -273,4 +276,40 @@ double LineSegment::distanceToLine(const LineSegment &line) const {
     return diff.length();
 }
 
+std::optional<LineSegment> LineSegment::shadow(const Vector2 &source, const LineSegment &obstacle, float negligible_shadow_length) const {
+    // If the source is on the obstacle line then the entire line is in the shadow, because every line from the source intersects with the obstacle line.
+    if (obstacle.isOnLine(source)) {
+        return length() > negligible_shadow_length ? std::optional(*this) : std::nullopt;
+    } else if (isPoint()) {
+        return std::nullopt;
+    }
+    std::optional<Vector2> firstIntersect = Line(*this).forwardIntersect(Line(source, obstacle.start));
+    std::optional<Vector2> secondIntersect = Line(*this).forwardIntersect(Line(source, obstacle.end));
+    if (!firstIntersect.has_value() && !secondIntersect.has_value()) {
+        return std::nullopt;
+    }
+    LineSegment shadow;
+    if (firstIntersect.has_value() && secondIntersect.has_value()) {
+        shadow = LineSegment(project(firstIntersect.value()), project(secondIntersect.value()));
+    } else {
+        Vector2 onlyIntersection = project(firstIntersect.has_value() ? firstIntersect.value() : secondIntersect.value());
+        double distanceIntersectionToStart = (start - onlyIntersection).length();
+        double distanceIntersectionToEnd = (end - onlyIntersection).length();
+        Vector2 closest = distanceIntersectionToStart < distanceIntersectionToEnd ? start : end;
+        Vector2 furthest = distanceIntersectionToStart < distanceIntersectionToEnd ? end : start;
+        bool furthestInShadow = LineSegment(source, furthest).doesIntersect(obstacle);
+        shadow = furthestInShadow ? LineSegment(onlyIntersection, furthest) : LineSegment(onlyIntersection, closest);
+    }
+    return shadow.length() > negligible_shadow_length ? std::optional(shadow) : std::nullopt;
```

Comment:
If you read the documentation then it is quite self-explanatory. However since you do not understand it directly, I will add additional comments that makes it more understandable.

Reply:
Ok so IMO there's 2 types of documentation and you should really make a clear distinction between the two.
1. User documentation which goes into the header (e.g. explaining what the function does, what it's arguments are and what it's limitations are), which is for people who are NOT interested in implementation details.
2. Code documentation. The occasional comment to explain what you are doing in complex calculations and why. I shouldn't have to look at the header file to understand what steps you are taking in this code here. This is what is missing here.

Reply:
Yes, I understand. This has already been updated...